File: same-tree/Solution100.java
// https://leetcode.com/problems/same-tree/

import java.util.LinkedList;
import java.util.Queue;

class Solution100 {

    // T: O(min(n,m)), S: O(min(h1,h2))
    public boolean isSameTree(TreeNode p, TreeNode q) {
        // If both nodes are null, trees are the same
        if (p == null && q == null) {
            return true;
        }

        // If one node is null while the other isn't, trees are different
        if (p == null || q == null) {
            return false;
        }

        // Check if current node values are the same
        if (p.val != q.val) {
            return false;
        }

        // Recursively check left and right subtrees
        return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
    }

    // T: O(min(n,m)), S: O(min(w1,w2))
    public boolean isSameTreeBfs(TreeNode p, TreeNode q) {
        // Use queues for BFS
        Queue<TreeNode> queue1 = new LinkedList<>();
        Queue<TreeNode> queue2 = new LinkedList<>();

        // Add root nodes to respective queues
        queue1.offer(p);
        queue2.offer(q);

        // BFS traversal
        while (!queue1.isEmpty() && !queue2.isEmpty()) {
            // Remove nodes from both queues
            TreeNode node1 = queue1.poll();
            TreeNode node2 = queue2.poll();

            // Check if both nodes are null
            if (node1 == null && node2 == null) {
                continue;
            }

            // If one node is null while other isn't, or values differ
            if (node1 == null || node2 == null || node1.val != node2.val) {
                return false;
            }

            // Add left children
            queue1.offer(node1.left);
            queue2.offer(node2.left);

            // Add right children
            queue1.offer(node1.right);
            queue2.offer(node2.right);
        }

        // Check if both queues are empty
        return queue1.isEmpty() && queue2.isEmpty();
    }

    public static void main(String[] args) {
        var sol = new Solution100();
        var p1 = new TreeNode(1);
        p1.left = new TreeNode(2);
        p1.right = new TreeNode(3);
        var q1 = new TreeNode(1);
        q1.left = new TreeNode(2);
        q1.right = new TreeNode(3);
        System.out.println(sol.isSameTree(p1, q1)); // true
        System.out.println(sol.isSameTreeBfs(p1, q1)); // true

        var p2 = new TreeNode(1);
        p2.left = new TreeNode(2);
        p2.right = null;
        var q2 = new TreeNode(1);
        q2.left = null;
        q2.right = new TreeNode(3);
        System.out.println(sol.isSameTree(p2, q2)); // false
        System.out.println(sol.isSameTreeBfs(p2, q2)); // false
    }

}

File: same-tree/solution100.go
// https://leetcode.com/problems/same-tree/

package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// T: O(min(n,m)), S: O(min(h1,h2))
func isSameTree(p *TreeNode, q *TreeNode) bool {
	if p == nil && q == nil {
		return true
	}

	if p == nil || q == nil {
		return false
	}

	if p.Val != q.Val {
		return false
	}

	return isSameTree(p.Left, q.Left) && isSameTree(p.Right, q.Right)
}

// T: O(min(n,m)), S: O(min(w1,w2))
func isSameTreeBfs(p *TreeNode, q *TreeNode) bool {
	queue1 := []*TreeNode{p}
	queue2 := []*TreeNode{q}

	for len(queue1) > 0 && len(queue2) > 0 {
		node1 := queue1[0]
		node2 := queue2[0]
		queue1 = queue1[1:]
		queue2 = queue2[1:]

		if node1 == nil && node2 == nil {
			continue
		}

		if node1 == nil || node2 == nil || node1.Val != node2.Val {
			return false
		}

		queue1 = append(queue1, node1.Left)
		queue2 = append(queue2, node2.Left)

		queue1 = append(queue1, node1.Right)
		queue2 = append(queue2, node2.Right)
	}

	return len(queue1) == 0 && len(queue2) == 0
}

func main() {
	p1 := &TreeNode{Val: 1}
	p1.Left = &TreeNode{Val: 2}
	p1.Right = &TreeNode{Val: 3}
	q1 := &TreeNode{Val: 1}
	q1.Left = &TreeNode{Val: 2}
	q1.Right = &TreeNode{Val: 3}
	fmt.Println(isSameTree(p1, q1))    // true
	fmt.Println(isSameTreeBfs(p1, q1)) // true

	p2 := &TreeNode{Val: 1}
	p2.Left = &TreeNode{Val: 2}
	p2.Right = nil
	q2 := &TreeNode{Val: 1}
	q2.Left = nil
	q2.Right = &TreeNode{Val: 3}
	fmt.Println(isSameTree(p2, q2))    // false
	fmt.Println(isSameTreeBfs(p2, q2)) // false
}
